Lib/typing.py

```diff
@@ -1406,6 +1406,12 @@ def _no_init_or_replace_init(self, *args, **kwargs):
     if cls._is_protocol:
         raise TypeError('Protocols cannot be instantiated')
 
+    # When `_no_init_or_replace_init` called using super() there are no
```

Comment:
Wait, is this only because of `super()`? I thought it's because once `__init__` is overriden and not using `_no_init_or_replace_init`, then there is no need to calculate?

I noticed Serhiy's original PR also had a guard that prevented calculation https://github.com/python/cpython/pull/27545/files#diff-ddb987fca5f5df0c9a2f5521ed687919d70bb3d64eaeb8021f98833a2a716887R1524.

Reply:
Basically, that because of `super()`.

I will try to explain using example:

```py
class P(Protocol):
    x: int


class Foo(P):
    def __init__(self):
        super().__init__()
```

`P.__init__` is `_no_init_or_replace_init` and `Foo.__init__` is `Foo.__init__`.

Previously, when we call `Foo.__init__` it called `P.__init__` (which is `_no_init_or_replace_init`) which search for a correct `__init__` and replace `Foo.__init__` with same `Foo.__init__`, after that it called resolved `__init__` which is `Foo.__init__` and it lead to `RecursionError`.

I hope I made things more clear.

Reply:
Thanks for the explanation Yurii.
I think we can shorten the comment, because the `super()` is slightly misleading. You can trigger with other ways:

```python
class P(Protocol):
    x: int


class Foo(P):
    def __init__(self):
        P.__init__(self)  # I know this is almost same as super(), but please bear with me :P
```
So the comment could say :
```
Already using a custom `__init__`. No need to calculate correct
`__init__` to call. This can lead to RecursionError. See bpo-45121.
```